### PR TITLE
Introduce Plugin Service Accounts 

### PR DESCRIPTION
**Pull request by scrawfor99**
### Description
This PR is part of addressing https://github.com/opensearch-project/security/issues/2941. 

It introduces the Application interface as well as a ServiceAccountManager interface which is implemented by  IdentityPlugins. With this new structure, when a plugin is installed, it is assigned a new ServiceAccount principal which can be used to reference the plugin. The service account principal is then assigned a set of permissions (to be implemented at a later date per https://github.com/opensearch-project/security/issues/2943) which it can use to operate without impersonating the undefined user context. 

Here is a diagram explaining the design idea: 

ServiceAccountsForPlugins(1)

### Related Issues

This PR is part of addressing https://github.com/opensearch-project/security/issues/2941. 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by scrawfor99** [comment]
By making an Application interface that extends the Subject interface, we run into an issue surrounding the treatment of principals and authentication. Because the Subject interface expects all subjects to implement getPrincipal() and authenticate(), we need to be able to forward this request to the IdentityPlugin. In turn, the IdentityPlugin can make use of the ApplicationManager and ServiceAccountManager to grab the associated principal and authenticate the application. However, this is a pretty complicated flow so I am not convinced there is not a better way of handling this.

File: server/src/main/java/org/opensearch/identity/ServiceAccount.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.List;
+
+/**
+ * This interface defines a ServiceAccount.
+ *
+ * A ServiceAccount is a principal which can retrieve its associated permissions.
+ */
+public interface ServiceAccount extends Principal {
+
+    public List<String> getPermissions();
```
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/ServiceAccount.java:21]: Working backwards, where would these permissions be defined? One of the primary reasons plugins stash the threadcontext (i.e. switching context to be able to do anything) is to write to a system index. What does the plugin/extension's relationship to the system index look like and how is it configured? Is it through `SystemIndex.getSystemIndexDescriptors` or is it configured elsewhere in a yml file?

Comment (scrawfor99) [reply server/src/main/java/org/opensearch/identity/ServiceAccount.java:21]: Hi @cwperks, great question! 

My initial thought for permission definition would be in the yml file for any plugins and defined in the extension installation request for extensions. The goal here is to eventually be able to offer some form of in-core FGAC as well as a clear boundary between what does and does not require impersonation. We will also be able to use the context switching to manage when we request tokens for authc/z. 

That being said, I am still working on the way I think permissions may look so this is not set in stone.

**Comment by scrawfor99** [comment]
Here is a diagram showing the proposed modified installation process with regard to the IndentityService.

ServiceAccountsInstallationFlow

**Comment by scrawfor99** [comment]
Here is a diagram of a request flow from an application (can be either a plugin or extension):

ServiceAccountsUseFlow

**Pull request closed without merging**